driver/src/driver/scheduler.rs

```diff
@@ -0,0 +1,247 @@
+use super::stablex_driver::StableXDriver;
+use crate::metrics::StableXMetrics;
+use crate::orderbook::StableXOrderBookReading;
+use anyhow::{anyhow, Context, Result};
+use log::error;
+use log::info;
+use std::collections::HashSet;
+use std::thread;
+use std::time::{Duration, Instant, SystemTime, SystemTimeError};
+
+const BATCH_DURATION: Duration = Duration::from_secs(300);
+
+#[derive(Clone, Copy, Debug, Eq, Hash, PartialEq)]
+struct BatchId(pub u64);
+
+impl BatchId {
+    pub fn current(now: SystemTime) -> std::result::Result<Self, SystemTimeError> {
+        let time_since_epoch = now.duration_since(SystemTime::UNIX_EPOCH)?;
+        Ok(Self(time_since_epoch.as_secs() / BATCH_DURATION.as_secs()))
+    }
+
+    pub fn currently_being_solved(now: SystemTime) -> std::result::Result<Self, SystemTimeError> {
+        Self::current(now).map(|batch_id| batch_id.prev())
+    }
+
+    pub fn start_time(self) -> SystemTime {
+        SystemTime::UNIX_EPOCH + Duration::from_secs(self.0 * BATCH_DURATION.as_secs())
+    }
+
+    pub fn solve_start_time(self) -> SystemTime {
+        self.start_time() + BATCH_DURATION
+    }
+
+    pub fn next(self) -> BatchId {
+        self.0.checked_add(1).map(BatchId).unwrap()
+    }
+
+    pub fn prev(self) -> BatchId {
+        self.0.checked_sub(1).map(BatchId).unwrap()
+    }
+}
+
+pub struct Scheduler<'a> {
+    driver: &'a mut dyn StableXDriver,
+    orderbook_reader: &'a dyn StableXOrderBookReading,
```

Comment:
We are not really reading the orderbook here anymore. May be worthwhile splitting this in to a OrderBookReading and AuctionStateReading at some point (maybe a separate PR)?

Reply:
I removed the get_auction_index call in the scheduler for simplicity of always waiting a fixed time.